File: software/examples/rust/example_callback.rs
use std::{error::Error, io, thread};
use tinkerforge::{imu_brick::*, ip_connection::IpConnection};

const HOST: &str = "localhost";
const PORT: u16 = 4223;
const UID: &str = "XXYYZZ"; // Change XXYYZZ to the UID of your IMU Brick.

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection.
    let imu = ImuBrick::new(UID, &ipcon); // Create device object.

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd.
                                          // Don't use device before ipcon is connected.

    // Create receiver for quaternion events.
    let quaternion_receiver = imu.get_quaternion_receiver();

    // Spawn thread to handle received events. This thread ends when the `imu` object
    // is dropped, so there is no need for manual cleanup.
    thread::spawn(move || {
        for quaternion in quaternion_receiver {
            println!("Quaternion [X]: {}", quaternion.x);
            println!("Quaternion [Y]: {}", quaternion.y);
            println!("Quaternion [Z]: {}", quaternion.z);
            println!("Quaternion [W]: {}", quaternion.w);
            println!();
        }
    });

    // Set period for quaternion receiver to 1s (1000ms).
    imu.set_quaternion_period(1000);

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}

File: software/examples/go/example_callback.go
package main

import (
	"fmt"
	"github.com/Tinkerforge/go-api-bindings/imu_brick"
	"github.com/Tinkerforge/go-api-bindings/ipconnection"
)

const ADDR string = "localhost:4223"
const UID string = "XXYYZZ" // Change XXYYZZ to the UID of your IMU Brick.

func main() {
	ipcon := ipconnection.New()
	defer ipcon.Close()
	imu, _ := imu_brick.New(UID, &ipcon) // Create device object.

	ipcon.Connect(ADDR) // Connect to brickd.
	defer ipcon.Disconnect()
	// Don't use device before ipcon is connected.

	imu.RegisterQuaternionCallback(func(x float32, y float32, z float32, w float32) {
		fmt.Printf("Quaternion [X]: %f\n", x)
		fmt.Printf("Quaternion [Y]: %f\n", y)
		fmt.Printf("Quaternion [Z]: %f\n", z)
		fmt.Printf("Quaternion [W]: %f\n", w)
		fmt.Println()
	})

	// Set period for quaternion receiver to 1s (1000ms).
	imu.SetQuaternionPeriod(1000)

	fmt.Print("Press enter to exit.")
	fmt.Scanln()

}
